File: mobile-engage/src/androidTest/java/com/emarsys/mobileengage/testUtil/ReflectionTestUtils.java
package com.emarsys.mobileengage.testUtil;

import java.lang.reflect.Constructor;
import java.lang.reflect.Field;

public class ReflectionTestUtils {

    public static void setStaticField(Class type, String fieldName, Object value) throws Exception {
        Field containerField = type.getDeclaredField(fieldName);
        containerField.setAccessible(true);
        containerField.set(null, value);
    }

    @SuppressWarnings("unchecked")
    public static <T> T getStaticField(Class type, String fieldName) throws Exception {
        Field field = type.getDeclaredField(fieldName);
        field.setAccessible(true);
        Object result = field.get(null);
        return (T) result;
    }

    @SuppressWarnings("unchecked")
    public static <T> T instantiate(Class type, int constructorIndex, Object... args) throws Exception {
        Constructor<?> constructor = type.getDeclaredConstructors()[constructorIndex];
        constructor.setAccessible(true);
        return (T) constructor.newInstance(args);
    }

}

File: testUtils/src/main/java/com/emarsys/test/util/ReflectionTestUtils.kt
package com.emarsys.test.util

object ReflectionTestUtils {

    @JvmStatic
    fun setStaticField(type: Class<*>, fieldName: String, value: Any?) {
        val containerField = type.getDeclaredField(fieldName)
        containerField.isAccessible = true
        containerField.set(null, value)
    }

    @JvmStatic
    fun <T> getStaticField(type: Class<*>, fieldName: String): T? {
        val field = type.getDeclaredField(fieldName)
        field.isAccessible = true
        val result = field.get(null)
        return result as T?
    }

    @JvmStatic
    fun <T> instantiate(type: Class<*>, constructorIndex: Int, vararg args: Any): T {
        val constructor = type.declaredConstructors[constructorIndex]
        constructor.isAccessible = true
        return constructor.newInstance(*args) as T
    }

}
